flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSource.java

```diff
@@ -96,21 +96,35 @@ public JdbcDynamicTableSource(
     public LookupRuntimeProvider getLookupRuntimeProvider(LookupContext context) {
         // JDBC only support non-nested look up keys
         String[] keyNames = new String[context.getKeys().length];
-        for (int i = 0; i < keyNames.length; i++) {
+
+        for (int i = 0; i < context.getKeys().length; i++) {
             int[] innerKeyArr = context.getKeys()[i];
             Preconditions.checkArgument(
                     innerKeyArr.length == 1, "JDBC only support non-nested look up keys");
             keyNames[i] = DataType.getFieldNames(physicalRowDataType).get(innerKeyArr[0]);
         }
+
         final RowType rowType = (RowType) physicalRowDataType.getLogicalType();
+
+        String[] conditions = null;
+        if (this.resolvedPredicates != null) {
+            conditions = new String[this.resolvedPredicates.size()];
+            for (int i = 0; i < this.resolvedPredicates.size(); i++) {
+                String resolvedPredicate = this.resolvedPredicates.get(i);
+                String param = this.pushdownParams[i].toString();
+                conditions[i] = (resolvedPredicate.replace("?", param));
```

Comment:
If you want to do it as Benchao Li suggested, ie. to use prepared statement. You can take a look at this class, its used in the ScanSource

https://github.com/apache/flink-connector-jdbc/blob/main/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcRowDataInputFormat.java#L159

Essentially it uses PrepareStatement to produce the final query string, from what I can tell `JdbcRowDataLookupFunction` already uses PreparedStatement, so we might be able to leverage that

Reply:
@qingwei91 Thanks for the pointer, at the point in the code that you are looking at we have access to the statement. But where I am in the code does not. I see that the JdbcFilterPushdownPreparedStatementVisitor hard codes a '?' literal as the placeholder for the parameter.

Reply:
@qingwei91 I have added a comment in the code detailing the alternatives I have looked into. Please let me know if I have missed something